.gitlab-ci.yml

```diff
@@ -0,0 +1,29 @@
+before_script:
+  - docker info
+  # check the availability of some commands:
+  - command >/dev/null -v python3 jq
+
+build-in-docker:
+  stage: build
+  script:
+    - cd tools/docker
+    # Try to pull from dockerhub, build the container images if that fails
+    - ./image-pull.sh || ./image-build.sh
+    # Run clang-format and check there are no modified files by checking git status
+    - ./clang-format.sh
```

Comment:
We can use the top-level clang-format.sh script itself if we run it in the appropriate docker container.

It would also be nice to put this in a separate job so it can be done in parallel (of course the git diff-index has to be included in the same job).

Again, both of these comments are for follow-up PRs.